www-src/app/replicator/design_documents.coffee

```diff
@@ -0,0 +1,151 @@
+async = require 'async'
+log = require('../lib/persistent_log')
+    prefix: "Design Documents"
+    date: true
+
+module.exports = class DesignDocuments
+
+    # Design names
+    @PATH_TO_BINARY: 'PathToBinary'
+    @FILES_AND_FOLDER: 'FilesAndFolder'
+    @PICTURES: 'Pictures'
+    @PHOTOS_BY_LOCAL_ID: 'PhotosByLocalId'
+    @BY_BINARY_ID: 'ByBinaryId'
+    @NOTIFICATIONS_TEMPORARY: 'NotificationsTemporary'
+    @LOCAL_PATH: 'LocalPath'
+    @CONTACTS: 'Contacts'
+    @CALENDARS: 'Calendars'
+
+    # Databases
+    cozyDB: null
+    internalDB: null
+
+    constructor: (@cozyDB, @internalDB) ->
+
+    createOrUpdateAllDesign: (callback) ->
+        log.info 'createOrUpdateAllDesign'
+        async.series [
+            (next) => @_createOrUpdate @cozyDB, \
+                DesignDocuments.NotificationsTemporaryDesignDoc, next
+            (next) => @_createOrUpdate @cozyDB, \
+                DesignDocuments.FilesAndFolderDesignDoc, next
+            (next) => @_createOrUpdate @cozyDB, \
+                DesignDocuments.ByBinaryIdDesignDoc, next
+            (next) => @_createOrUpdate @cozyDB, \
+                DesignDocuments.PicturesDesignDoc, next
+            (next) => @_createOrUpdate @cozyDB, \
+                DesignDocuments.LocalPathDesignDoc, next
+            (next) => @_createOrUpdate @cozyDB, \
+                DesignDocuments.PathToBinaryDesignDoc, next
+            (next) => @_createOrUpdate @cozyDB, \
+                DesignDocuments.ContactsDesignDoc, next
+            (next) => @_createOrUpdate @cozyDB, \
+                DesignDocuments.CalendarsDesignDoc, next
+            (next) => @_createOrUpdate @internalDB, \
+                DesignDocuments.PhotosByLocalIdDesignDoc, next
+        ], callback
+
+    _createOrUpdate: (db, design, callback) ->
+        db.get design._id, (err, existing) =>
+            if existing?.version is design.version
+                callback null, {}
+            else
+                log.info "_createOrUpdate: #{design._id} FROM #{existing}"
+                design._rev = existing._rev if existing
+                db.put design, callback
+
+
+    # Design Documents
+
+    @PathToBinaryDesignDoc:
+        _id: "_design/#{@PATH_TO_BINARY}"
+        version: 1
+        views:
+            "#{@PATH_TO_BINARY}":
+                map: Object.toString.apply (doc) ->
+                    if doc.docType?.toLowerCase() is 'file'
+                        emit doc.path + '/' + doc.name, doc.binary?.file?.id
+
+
+    @FilesAndFolderDesignDoc:
+        _id: "_design/#{@FILES_AND_FOLDER}"
+        version: 1
+        views:
+            "#{@FILES_AND_FOLDER}":
+                map: Object.toString.apply (doc) ->
+                    if doc.name?
+                        if doc.docType?.toLowerCase() is 'file'
+                            emit [doc.path, '2_' + doc.name.toLowerCase()]
+                        if doc.docType?.toLowerCase() is 'folder'
```

Comment:
@kosssi you should probably add a comment for this view that explain _why_ it forces folders to come before files.